agent/whatweb.py

```diff
@@ -0,0 +1,135 @@
+                       'X-XSS-Protection', 'x-pingback', 'Strict-Transport-Security',
+                       'UncommonHeaders', 'HTML5', 'Script', 'Title', 'Email', 'Meta-Author',
+                       'Frame', 'PasswordField', 'MetaGenerator', 'Object', 'Country', 'IP']
+
+DEFAULT_FINGERPRINT = 'BACKEND_COMPONENT'
+
+FINGERPRINT_TYPE = {
+    'jquery': 'JAVASCRIPT_LIBRARY'
+}
+
+WHATWEB_PATH = './whatweb'
+WHATWEB_DIRECTORY = '/WhatWeb'
+SELECTOR = 'v3.fingerprint.domain_name.library'
+
+
+class WhatWebAgent(agent.Agent):
+    """Agent responsible for finger-printing a website."""
+
+    def _start_scan(self, domain_name: str, output_file: Union[str, bytes, os.PathLike]):
+        """Run a whatweb scan using python subprocess.
+
+        Args:
+            target:  Target
+            output_file: The output file to save the scan result.
+        """
+        logger.info('Staring a new scan for %s .', domain_name)
+        whatweb_command = [WHATWEB_PATH, f'--log-json-verbose={output_file}', domain_name]
+        subprocess.run(whatweb_command, cwd=WHATWEB_DIRECTORY, check=True)
+
+    def _parse_result(self, domain_name: str, output_file: io.BytesIO):
+        """After the scan is done, parse the output json file into a dict of the scan findings."""
+        try:
+            # whatweb writes duplicate lines in some cases, breaking json. We process only the first line.
+            file_contents = output_file.readlines()
+            if file_contents is not None and len(file_contents) > 0:
+                for file_content in file_contents:
+                    results = json.loads(file_content)
+                    for result in results:
+                        if isinstance(result, list):
+                            for list_plugin in result:
+                                if len(list_plugin) > 0:
+                                    plugin = list_plugin[0]
+                                else:
+                                    plugin = list_plugin
+                                if plugin not in BLACKLISTED_PLUGINS:
+                                    values = list_plugin[1]
+                                    versions = ''
+                                    name = plugin
+                                    for value in values:
+                                        if 'version' in value:
+                                            versions = value['version']
+                                        if 'string' in value:
+                                            name = str(value['string'])
+                                    self._send_detected_fingerprints(domain_name, name, versions)
+                logger.info('Scan is done Parsing the results from %s.', output_file.name)
+        except OSError as e:
+            logger.error('Exception while processing %s with message %s', output_file, e)
+
+    def _send_detected_fingerprints(self, domain_name: str, name: str, versions: Union[list, str]):
```

Comment:
This method should always get passed versins as a list. If we have only, put it in a list.